libcyphal_demo/src/platform/block_memory_resource.hpp

```diff
@@ -0,0 +1,165 @@
+// This software is distributed under the terms of the MIT License.
+// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
+// Copyright Amazon.com Inc. or its affiliates.
+// SPDX-License-Identifier: MIT
+// Author: Sergei Shirokov <sergei.shirokov@zubax.com>
+
+#ifndef PLATFORM_BLOCK_MEMORY_RESOURCE_HPP
+#define PLATFORM_BLOCK_MEMORY_RESOURCE_HPP
+
+#include <cetl/cetl.hpp>
+#include <cetl/pf17/cetlpf.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+
+namespace platform
+{
+
+/// Implements a C++17 PMR memory resource that uses a pool of pre-allocated blocks.
+///
+template <std::size_t MaxAlign = sizeof(std::max_align_t)>
+class BlockMemoryResource : public cetl::pmr::memory_resource
+{
+public:
+    struct Diagnostics final
+    {
+        std::size_t   capacity;
+        std::size_t   allocated;
+        std::size_t   peak_allocated;
+        std::size_t   block_size;
+        std::uint64_t oom_count;
+
+    };  // Diagnostics
+
+    BlockMemoryResource()           = default;
+    ~BlockMemoryResource() override = default;
+
+    BlockMemoryResource(BlockMemoryResource&&)                 = delete;
+    BlockMemoryResource(const BlockMemoryResource&)            = delete;
+    BlockMemoryResource& operator=(BlockMemoryResource&&)      = delete;
+    BlockMemoryResource& operator=(const BlockMemoryResource&) = delete;
+
+    void setup(const std::size_t pool_size_bytes, void* const pool, const std::size_t block_size_bytes)
```

Comment:
Why is this not done in the constructor? Post-initialization is almost always a bad idea because it makes it possible for an obect to reside in an invalid state, and there are new edge cases when post-init is done a number of times other than one.